Libertydevops 1 gitjenkinshelmiks (#2245)
OpenLiberty/blogs: posts/2022-05-27-libertydevops-cicd-jenkins.adoc

---
layout: post
title: "Continuous integration and delivery (CI/CD) of cloud-native Java applications using Jenkins"
categories: blog
author_picture: https://avatars3.githubusercontent.com/shamjithantholi
author_github: https://github.com/shamjithantholi
seo-title: Continuous integration and delivery (CI/CD) of cloud-native Java applications using Jenkins - OpenLiberty.io
seo-description: How to build a DevOps pipeline for a cloud-native Java application with Jenkins as the core controller component.
blog_description: How to build a DevOps pipeline for a cloud-native Java application with Jenkins as the core controller component.
open-graph-image: https://openliberty.io/img/blog/liberty-devops-generic-architecture.png
---
= Continuous integration and delivery (CI/CD) of cloud-native Java applications using Jenkins
Shamjith Antholi <https://github.com/shamjithantholi>
:imagesdir: /
:url-prefix:
:url-about: /

[#Intro]

> Continuous integration and continuous deployment (CI/CD) is a process in which developers frequently merge code changes into a source code repository, which then triggers code builds, runs security scans, runs tests, and then deploys the application. Automation of this process, through adopting a DevOps culture, tools, and processes, enables organizations to release on a more frequent basis without compromising on quality. There are various tools available to automate the CI/CD process that can be readily integrated into a DevOps deployment pipeline. In this blog post we'll discuss how you can use link:https://www.jenkins.io/[Jenkins] to automate the CI/CD process for building, testing, and deploying your cloud-native Java applications running on link:https://openliberty.io[Open Liberty]. As the CI/CD process is triggered by source changes in Git, this is also known as the GitOps approach to DevOps.

Reviewer: ```suggestion
Continuous integration and continuous deployment (CI/CD) is a process in which developers frequently merge code changes into a source code repository, which then triggers code builds, runs security scans, runs tests, and then deploys the application. Automation of this process, through adopting a DevOps culture, tools, and processes, enables organizations to release on a more frequent basis without compromising on quality. Various tools are available to automate the CI/CD process that can be readily integrated into a DevOps deployment pipeline. In this blog post, we'll discuss how you can use link:https://www.jenkins.io/[Jenkins] to automate the CI/CD process for building, testing, and deploying your cloud-native Java applications that run on link:https://openliberty.io[Open Liberty]. As the CI/CD process is triggered by source changes in Git, this is also known as the GitOps approach to DevOps.
```

> Jenkins is a popular open source automation server that has hundreds of plugins to help integrate it into your existing DevOps pipeline. The following diagram shows a simple architecture for a Jenkins deployment pipeline that builds and deploys a Jakarta EE and MicroProfile application in a single Kubernetes environment. When a code change is made in the Git repository, GitHub will trigger link:https://www.jenkins.io[Jenkins] to build the code, runs JUnit tests and then link:https://www.sonarqube.org[SonarQube] static code analysis scans against the built code. Jenkins then generates a link:https://www.docker.com[Docker] container image of the application with Open Liberty as its runtime, saves the image to a Docker container repository and then scans the image with link:https://www.aquasec.com[Trivy]. Jenkins then triggers a deployment using CLI commands and deploys the Docker container image to link:https://kubernetes.io[Kubernetes].

Reviewer: ```suggestion
Jenkins is a popular open source automation server that has hundreds of plug-ins to help integrate it into your existing DevOps pipeline. The following diagram shows a simple architecture for a Jenkins deployment pipeline that builds and deploys a Jakarta EE and MicroProfile application in a single Kubernetes environment. When a code change is made in the Git repository, GitHub will trigger link:https://www.jenkins.io[Jenkins] to build the code, runs JUnit tests and then link:https://www.sonarqube.org[SonarQube] static code analysis scans against the built code. Jenkins then generates a link:https://www.docker.com[Docker] container image of the application with Open Liberty as its runtime, saves the image to a Docker container repository, and then scans the image with link:https://www.aquasec.com[Trivy]. Jenkins then triggers a deployment by using CLI commands and deploys the Docker container image to link:https://kubernetes.io[Kubernetes].
```

Author: Done

                         docker build -t liberty-$<code identifier>:$<docker image version> .
                         #eg: docker build -t liberty-app:v1.0 .
                         docker tag liberty-$<code identifier>:$<docker image version> <remote-docker-image-repository-url>/<docker-repo-name>/liberty-$<code identifier>:$<docker image version>
                         #eg: docker tag liberty-app:v1.0 docker.io/someid/liberty-app:v1.0
                         docker push <remote-docker-image-repository-url>/<docker-repo-name>/liberty-$<code identifier>:$<docker image version>
                         #eg: docker push docker.io/someid/liberty-app:v1.0  
                       '''   
                  }
               }
            }
          }
----

Where:

* _git-token_ is the personal access token you have generated from your Github account.
* _remote-docker-image-repository-url_ is the location of the Docker image repository.
* _username_ is your user name for the Docker image repository.
* _password_ is your password for the Docker image repository.
* _docker-image-version_ is the version number of your Docker image, a unique identifier.

> Build your container image using a Dockerfile or a Containerfile generated by link:http://start.openliberty.io[Liberty Starter] or by following the pattern described at link:https://github.com/OpenLiberty/ci.docker[here]. The link:https://openliberty.io/guides/containerize.html[Containerize guide] is a helpful resource that goes into more details on how to create a container image for applications running on Liberty.

Reviewer: ```suggestion
Build your container image using a Dockerfile or a Containerfile generated by the link:http://start.openliberty.io[Liberty Starter] or by following the pattern described in link:https://github.com/OpenLiberty/ci.docker[Open Liberty Images]. The link:https://openliberty.io/guides/containerize.html[Containerize guide] is a helpful resource that goes into more details on how to create a container image for applications running on Liberty.
```

For static code analysis, you can use SonarQube Community edition. The Jenkins client setup details for SonarQube are given at link:https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-jenkins/[SonarScanner for Jenkins]. The following sample Maven command packages the code with Maven and runs the SonarQube scan:

[source]
----
mvn package verify sonar:sonar -Dsonar.projectKey=sampleapp -Dsonar.host.url=http://localhost:9000 -Dsonar.login=<generated-login-key>
----

For scanning the container image (for security purpose), you can use link:https://hub.docker.com/r/aquasec/trivy/[Trivy]. This scan provides the vulnerability details of open source JAR files that are used to resolve the dependencies while building the application. The following Docker command runs Trivy against your container image:

[source]
----
docker run aquasec/trivy image docker.io/<docker-repo>/liberty-app:v1.0
----

Where:

* _docker-repo_ is the name of the Docker container repository that contains your image

== Deploying cloud-native Java applications with Open Liberty to Kubernetes with Jenkins

> For simplicity, I will use the command line (CLI) code in Jenkins pipeline job to deploy a Jakarta EE and MicroProfile application with Open Liberty into Kubernetes. You can also use other tools like Helm, Travis CI, and CircleCI.

Reviewer: ```suggestion
For simplicity, I will use the command line (CLI) code in a Jenkins pipeline job to deploy a Jakarta EE and MicroProfile application with Open Liberty into Kubernetes. You can also use other tools like Helm, Travis CI, and CircleCI.
```

kubectl create -f deploy/deployment.yaml #( simple k8s deployment command )
kubectl create -f deploy/service.yaml #( simple k8s service creation command )
kubectl create -f deploy/route.yaml #( simple k8s route creation command )
----

Make sure that your Kubernetes configuration files are stored in the same Git repository as your Jenkinsfile in a sub-directory called `deploy`. Also ensure that the Docker image name in the Kubernetes deployment configuration file is updated according to the container image name/tag in the Dockerfile (manually, or programmatically if it needs to change at run time):

[.imageblock.img_border_light]
image::/img/blog/deploymentyaml.png[Image reference in deployment yaml,width=40%,align="center"]

When Jenkins has checked out the Java application code for the code build, all the Kubernetes configuration files are also downloaded to the Jenkins workspace so that Jenkins can run the IBM Cloud and Kubernetes commands to connect to the Kubernetes cluster and deploy the application.

See the link:https://kubernetes.io/docs/reference/kubectl/cheatsheet/[Kubernetes documentation] for other commands.

== QA testing cloud-native Java applications with Jenkins

> Apart from running JUnit test cases along with the code build phase, Jenkins can trigger functional and integration QA test cases automatically after deploying the cloud-native Java application.

Reviewer: ```suggestion
Apart from running JUnit test cases along with the code build phase, Jenkins can automatically trigger functional and integration QA test cases after deploying the cloud-native Java application.
```

Configure the test cases in the Jenkins job and test it manually. Create a remote job identifier authentication token in the "Trigger builds remotely" section under "Build Triggers". Trigger this test case from the Docker "entrypoint" file by using a remote rest API call that uses this authentication token as the identifier.

For example, run the following command in a terminal: 

[source]
----
curl -I -u <auth-token> https://<jenkins-host>/job/<job-name>/build?token=<remote-job-identifier-authentication-token>
----

You can generate an authentication token (auth-token) with link:https://www.postman.com[Postman] using the Jenkins login credentials.

== Kubernetes monitoring tools

> Kubernetes provides commands to check the application or cluster logs and memory and CPU usage through the commands like: 

Reviewer: ```suggestion
You can use the following Kubernetes commands to check the application or cluster logs and the memory and CPU usage:
```

[source]
----
kubectl logs ..
cat /sys/fs/cgroup/cpu/cpuacct.usage (after connecting to k8s pod)
cat /sys/fs/cgroup/memory/memory.usage_in_bytes (after connecting to k8s pod)
----

> There are several applications available that you can integrate with Kubernetes to persist logs and usage statistics, such as link:https://prometheus.io[Prometheus] and link:https://grafana.com/oss/loki/[Grafana].

Reviewer: ```suggestion
You can integrate different applications with Kubernetes to persist logs and usage statistics, such as link:https://prometheus.io[Prometheus] and link:https://grafana.com/oss/loki/[Grafana].
```

> Liberty makes it easy for system and application metrics to be collected and visualized for observability using Prometheus and Grafana. You can find guidance and more details in the resources listed link:https://community.ibm.com/community/user/wasdevops/blogs/don-bourne1/2021/06/26/metrics-and-monitoring-guidance-for-open-liberty-a[here].

Reviewer: ```suggestion
Liberty makes it easy to collect and visualize system and application metrics for observability by using Prometheus and Grafana. You can find guidance and more details in the resources listed link:https://community.ibm.com/community/user/wasdevops/blogs/don-bourne1/2021/06/26/metrics-and-monitoring-guidance-for-open-liberty-a[here].
```

== Conclusion

> There are many ways in which you can configure your DevOps pipeline. This blog post is a quick introduction to how you can use Jenkins to set up a simple CI/CD pipeline to build and deploy your cloud-native Java applications on Liberty.

Reviewer: ```suggestion
You can configure your DevOps pipeline in many ways. This blog post is a quick introduction to how you can use Jenkins to set up a simple CI/CD pipeline to build and deploy your cloud-native Java applications on Liberty.
```
